doc/source/rllib/single-agent-episode.rst

```diff
@@ -0,0 +1,214 @@
+    `actions`, `rewards`, `infos`, and `extra_model_outputs`. For simplicity, only the getters for observations, actions, and rewards
+    are shown here. Their behavior is intuitive, returning a single item when provided with a single index and returning a list of items
+    (in the non-finalized case; see further below) when provided with a list of indices or a slice (range) of indices.
+
+
+The following code snippet summarizes the various capabilities of the different getter methods:
+
+.. literalinclude:: doc_code/sa_episode.py
+    :language: python
+    :start-after: rllib-sa-episode-02-begin
+    :end-before: rllib-sa-episode-02-end
+
+
+Finalized and Non-Finalized Episodes
+------------------------------------
+
+The data in a :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` can exist in two states:
+non-finalized and finalized. A non-finalized episode stores its observations and other data items in
+plain python lists and appends new timestep data to these. In a finalized episode,
+these lists have been converted into (possibly complex) structures that have NumPy arrays at their leafs.
+Note that a "finalized" episode doesn't necessarily have to be terminated or truncated yet
+in the sense that the underlying RL environment declared the episode to be over (or has reached some
+maximum number of timesteps).
+
+.. figure:: images/episodes/sa_episode_non_finalized_vs_finalized.svg
+    :width: 900
+    :align: left
+
+
+:py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` objects start in the non-finalized
+state (data stored in python lists), making it very fast to append data from an ongoing episode:
+
+
+.. literalinclude:: doc_code/sa_episode.py
+    :language: python
+    :start-after: rllib-sa-episode-03-begin
+    :end-before: rllib-sa-episode-03-end
+
+
+To illustrate the differences between the data stored in a non-finalized episode vs the same data stored in
+a finalized one, take a look at this complex observation example here, showing the exact same observation data in two
+episodes (one not finalized the other finalized):
+
+.. figure:: images/episodes/sa_episode_non_finalized.svg
+    :width: 800
+    :align: left
+
+    **Complex observations in a non-finalized episode**: Each individual observation is a (complex) dict matching the
+    gym environment's observation space. There are three such observation items stored in the episode so far.
+
+.. figure:: images/episodes/sa_episode_finalized.svg
+    :width: 600
+    :align: left
+
+    **Complex observations in a finalized episode**: The entire observation record is a single (complex) dict matching the
+    gym environment's observation space. At the leafs of the structure are `np.NDArrays` holding the individual values of the leaf.
+    Note that these `NDArrays` have an extra batch dim (axis=0), whose length matches the length of the episode stored (here 3).
```

Comment:
Either both `np.NDArray`  or both `NDArray` I'd say. What do you think?

Reply:
fixed